File: MovieDB-Java/app/src/main/java/com/practice/moviedb/activities/MainActivity.java
package com.practice.moviedb.activities;

import android.os.Bundle;

import com.practice.moviedb.R;
import com.practice.moviedb.adapters.MovieListAdapter;
import com.practice.moviedb.databinding.ActivityMainBinding;
import com.practice.moviedb.models.TopRatedMovie;
import com.practice.moviedb.networks.ApiClient;
import com.practice.moviedb.networks.ApiService;
import com.practice.moviedb.repositories.TopRatedMovieRepository;
import com.practice.moviedb.viewmodels.TopRatedMovieViewModel;
import com.practice.moviedb.viewmodels.factories.TopRatedMovieVMFactory;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;


public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding binding;

    private TopRatedMovieViewModel viewModel;

    private MovieListAdapter adapter;

    private ProgressBar progressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        setSupportActionBar(binding.toolbar);

        progressBar = findViewById(R.id.progress_bar);

        initViewModel();
        initRecyclerView();
        dataObserver();
    }

    private void initViewModel() {

        progressBar.setVisibility(View.VISIBLE);

        ApiService apiService = ApiClient.getClient().create(ApiService.class);
        TopRatedMovieRepository repository = new TopRatedMovieRepository(apiService);

        viewModel = ViewModelProviders.of(this, new TopRatedMovieVMFactory(repository))
                .get(TopRatedMovieViewModel.class);

        viewModel.initTopRatedMovieFromRepo(
                "6371db70ffc8e719f981e307e397452e",
                "en-US",
                "1",
                "vote_average.asc");

        binding.setTopRatedMovieModel(viewModel);
    }

    private void dataObserver() {
        viewModel.getTopRatedMovieLiveData().observe(this, new Observer<TopRatedMovie>() {
            @Override
            public void onChanged(TopRatedMovie topRatedMovie) {

                progressBar.setVisibility(View.GONE);

                updateRecyclerView(topRatedMovie);
            }
        });
    }

    private void updateRecyclerView(TopRatedMovie topRatedMovie) {
        if (topRatedMovie != null) {
            adapter.setTopRatedMovie(topRatedMovie);
            adapter.notifyDataSetChanged();
        }
    }

    private void initRecyclerView() {

        // Initializing with blank list

        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        binding.containMain.movieList.setLayoutManager(layoutManager);

        adapter = new MovieListAdapter(this);
        adapter.setTopRatedMovie(new TopRatedMovie());
        binding.containMain.movieList.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: MovieDB-Kotlin/app/src/main/java/com/practice/moviedatabase/activities/MainActivity.kt
package com.practice.moviedatabase.activities

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.practice.moviedatabase.R
import com.practice.moviedatabase.AllConstants
import com.practice.moviedatabase.adapters.MovieListAdapter
import com.practice.moviedatabase.base.ItemClickListener
import com.practice.moviedatabase.models.Result
import com.practice.moviedatabase.models.TopRatedMovie
import com.practice.moviedatabase.models.params.TopRatedMovieParams
import com.practice.moviedatabase.networks.ApiClient
import com.practice.moviedatabase.networks.ApiService
import com.practice.moviedatabase.repositories.TopRatedMovieRepository
import com.practice.moviedatabase.viewmodels.TopRatedMovieViewModel
import com.practice.moviedatabase.viewmodels.factories.TopRatedViewModelFactory

import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*

class MainActivity : AppCompatActivity(), ItemClickListener {

    private lateinit var viewModel : TopRatedMovieViewModel
    private lateinit var adapter : MovieListAdapter

    private var checked: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)

        initViewModel()
        initViews()
        dataObserver()
    }

    private fun dataObserver() {
        viewModel.topRateMovieLiveData.observe(this, Observer {
            if (it != null) {
                progressBar.visibility = View.GONE
                adapter.setTopRatedMovie(it)
                adapter.notifyDataSetChanged()
            }
        })
    }

    private fun initViews() {

        adapter = MovieListAdapter(this)
        adapter.setTopRatedMovie(TopRatedMovie())
        adapter.setItemClickListener(this)

        val layoutManager = LinearLayoutManager(this)
        movieListRecyclerView.layoutManager = layoutManager
        movieListRecyclerView.adapter = adapter

        detailsViewSwitch.isChecked = false

        detailsViewSwitch.setOnCheckedChangeListener { _, isChecked ->
            checked = isChecked
        }
    }

    private fun initViewModel() {
        progressBar.visibility = View.VISIBLE

        val apiService = ApiClient.client?.create(ApiService::class.java)
        val repository = TopRatedMovieRepository(apiService)

        viewModel = ViewModelProviders.of(this, TopRatedViewModelFactory(repository))
            .get(TopRatedMovieViewModel::class.java)

        viewModel.requestTopRatedMoviesApi(TopRatedMovieParams(getString(R.string.api_key), getString(R.string.language),
            getString(R.string.default_page), getString(R.string.sorted_by)))
    }

    override fun onItemClicked(result: Result, outputDate: String?) {

        val intent = Intent(this, MovieDetailsActivity::class.java)
        intent.putExtra("poster_url", AllConstants.BASE_IMAGE_URL + result.posterPath)
        intent.putExtra("title", result.title)
        intent.putExtra("release_date", outputDate)
        intent.putExtra("vote_average", result.voteAverage.toString())
        intent.putExtra("overview", result.overview)
        intent.putExtra("checked", checked)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return when (item.itemId) {
            R.id.action_settings -> true
            else -> super.onOptionsItemSelected(item)
        }
    }
}
